src/libraries/System.Net.Security/src/System/Net/Security/SslConnectionInfo.Unix.tt

```diff
@@ -0,0 +1,89 @@
+﻿<#@ template debug="false" hostspecific="false" language="C#" #>
+<#@ import namespace="System" #>
+<#@ assembly name="System.Core" #>
+<#@ assembly name="System.Net.Primitives" #>
+<#@ import namespace="System.Linq" #>
+<#@ import namespace="System.Text" #>
+<#@ import namespace="System.Collections.Generic" #>
+<#@ output extension=".cs" #>
+<#@ include file="TlsCipherSuiteNameParser.ttinclude" #><#@
+    include file="TlsCipherSuite.cs" #>
+<# Array enumValues = typeof(TlsCipherSuite).GetEnumValues(); #>
+
+
+using System.Diagnostics;
+using System.Security.Authentication;
+
+namespace System.Net.Security
+{
+    internal partial struct SslConnectionInfo
+    {
+        private void MapCipherSuite(TlsCipherSuite cipherSuite)
+        {
+            TlsCipherSuite = cipherSuite;
+            KeyExchKeySize = 0;
+
+            int index = (int)cipherSuite * 2;
+            Debug.Assert(s_cipherToIndex.IndexOf(index) > -1, $"No mapping found for cipherSuite {cipherSuite}");
+            ulong data = s_encodedNumber[s_cipherToIndex[index + 1]];
```

Comment:
`ulong` is 4 bytes more than `int`. We have ~300 cases, so encoding the raw data in more compact form as ints is going to be about 1.2kB smaller.

The code to decode a more compact form can be a bit bigger, but it is going to be a lot less than 1.2kB.

Reply:
I've divided the switch-case helper into two. We are saving about 2 KB on NativeAOT but losing about 1 KB in assembly size; which is still 14.5 KB better than the `main` branch.

Reply:
It should be possible to encode the data into a single `int`: https://github.com/dotnet/runtime/pull/84690#discussion_r1167142258 . Thoughts about this approach? I would expect it to be the most compact.

Reply:
I'm still experimenting with your approach and pushed some changes. It is showing up best results so far. 👍

It is a bit unfortunate that value of `DataKeySize` ranges from 0-256 both bounds inclusive, so I had to pack them like this:

| 4 bytes | 4 bytes | 16 bytes | 8 bytes |
| --- | --- | --- | --- |
| ExchangeAlgorithmTypeIndex | CipherAlgorithmTypeIndex | CipherAlgorithmStrength | HashAlgorithmTypeIndex |

Reply:
The interesting threshold is packing the data down to 16 bits. It would make a difference on arm64. Loading immediate that fits into 16 bits is one instruction on arm64. Loading immediate that has more than 16 bits is two instructions on arm64.

If you use more than 16 bits for the data encoding, it does not really matter whether it is 17 or all 32 bits. The instruction encoding is going to have about the same code size.

> DataKeySize ranges from 0-256 both bounds inclusive, 

There are only a few values used in the range, so you can use the same `Index` approach as for the rest.

Reply:
16 bits version on osx-arm64 has increased the assembly size: `252928` bytes with 32 bits, now: `253440` bytes as well as the NativeAOT app; crossgen2 binary was: `28474696` bytes with 32 bits, now: `28475144` bytes. If there is still a benefit of keeping the 16 bit version, then we can ignore this minor difference; as compared to main branch, overall diff of NativeAOT > 32 K in both versions. Otherwise, I can revert back to 32 bit version: https://github.com/dotnet/runtime/pull/84690/commits/ea293449913334379991c0bb5da0348c22e751f1.

Reply:
> 16 bits version on osx-arm64 has increased the assembly size

That's odd. It is not what I would expect. Did you happen to look at the generate code to understand what makes the 16-bit version larger? Maybe RyuJIT has a codegen problem with `ushort`? If it is the case, it should be possible to fix it by changing the return the type `int`.

Either way, this looks fine - the difference is not that big.

Reply:
I didn't check the cogegen but you are right about the `ushort`, changing the return type to int recovered the extra bytes and more. Now we are at `28474632` with crossgen2 binary on osx-arm64. 👍

Reply:
I didn't check the codegen but you are right about the `ushort`, changing the return type to int removed the extra bytes and more. Now we are at `28474632` with crossgen2 binary on osx-arm64. 👍

Reply:
OK less try it now